### PR TITLE
002 removing authentication

### PR DESCRIPTION
**Pull request by amcmurray-bw**
- added scheduled task to get random tweets every 15 mins
- removed twitter verification 
- configuration file added

### CODE REVIEW

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetController.java [hunk 1]
```diff
@@ -1,37 +1,21 @@
...
-import org.springframework.social.twitter.api.Twitter;
 import org.springframework.stereotype.Controller;
-import org.springframework.ui.Model;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.GetMapping;
 
 @Controller
-@RequestMapping("/")
 public class TweetController {
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetController.java]: Is there still a need for this controller at all?

File: twitter-crawler/pom.xml [hunk 2]
```diff
@@ -19,25 +19,33 @@
...
         </dependency>
+
         <dependency>
             <groupId>org.springframework.social</groupId>
             <artifactId>spring-social-twitter</artifactId>
         </dependency>
+
+        <dependency>
+            <groupId>org.springframework.social</groupId>
+            <artifactId>spring-social-web</artifactId>
```
Comment (YNedderhoff) [inline twitter-crawler/pom.xml]: Where's this used? Do you need both `spring-social-twitter` and `spring-social-web`?

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java [hunk 3]
```diff
@@ -9,21 +13,34 @@
...
 @Service
 public class TweetService {
 
     private final TweetRepository tweetRepository;
     private final Twitter twitter;
+    private final ExecutorService scheduledTaskExecutorService;
 
     @Autowired
-    public TweetService(TweetRepository tweetRepository, Twitter twitter) {
+    public TweetService(TweetRepository tweetRepository, Twitter twitter, ExecutorService scheduledTaskExecutorService) {
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: consider putting every constructor param into a new line to avoid lengthy lines

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java [hunk 4]
```diff
@@ -9,21 +13,34 @@
...
         this.tweetRepository = tweetRepository;
         this.twitter = twitter;
+        this.scheduledTaskExecutorService = scheduledTaskExecutorService;
     }
 
+    //cron set to every 15 minutes on the hour, eg 12:00, 12:15 etc
+    @Scheduled(cron = "0 0/15 * * * *")
     public void getTweetsAndSaveToDB() {
+        System.out.println("Scheduled task");
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: you probably don't want these anymore, and if you do, maybe as a log on `DEBUG` level?

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java [hunk 5]
```diff
@@ -9,21 +13,34 @@
...
+        CompletableFuture.runAsync(() -> saveTweetsIntoDB(), scheduledTaskExecutorService).exceptionally(throwable -> handleAsyncException(throwable));
+    }
+
+    public Void handleAsyncException(Throwable throwable) {
+        System.out.println(throwable.getCause());
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: You are going to want to replace this with a sensible log (i.e. a message + stack trace or so)

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java [hunk 6]
```diff
@@ -33,6 +50,7 @@ public void getTweetsAndSaveToDB() {
             SavedTweet savedTweet = new SavedTweet(tweet.getId(), tweet.getText(), "-1");
             tweetRepository.insert(savedTweet);
         }
-    }
 
+        System.out.println("Tweets saved");
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: Same here: logging

File: twitter-crawler/src/main/resources/templates/hello.html [hunk 7]
```diff
@@ -4,9 +4,9 @@
 
 </head>
 <body>
-<h3>Hello, <span th:text="${twitterProfile.name}">Some User</span>!</h3>
+<h3>Hello!</h3>
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/resources/templates/hello.html]: Is this file still needed at all?

File: .gitignore [hunk 8]
```diff
@@ -1,4 +1,6 @@
 /target/
+twitter-crawler/target/
```
Comment (YNedderhoff) [inline .gitignore]: I wonder if you replace `/target/` with `target/` if it works for all target folders in the tree?

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java [hunk 9]
```diff
@@ -9,21 +16,43 @@
...
         this.tweetRepository = tweetRepository;
         this.twitter = twitter;
+        this.scheduledTaskExecutorService = scheduledTaskExecutorService;
     }
 
+    //cron set to every 15 minutes on the hour, eg 12:00, 12:15 etc
+    @Scheduled(cron = "0 0/15 * * * *")
     public void getTweetsAndSaveToDB() {
 
+        logger.info("Scheduled task started at " + new Date());
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: Two things. Firstly, I think you should be able to replace this with 

` logger.info("Scheduled task started at {}", new Date());`

The `{}` syntax comes in handy especially when you have multiple placeholders, but it's good practice in general.

Second, I wouldn't use the old `Date` but rather the new Time API, for example `LocalDateTime.now()` or so.

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java [hunk 10]
```diff
@@ -9,21 +16,43 @@
...
+        CompletableFuture.runAsync(() -> saveRandomTweetsIntoDB(),
+                scheduledTaskExecutorService).exceptionally(throwable -> handleAsyncException(throwable));
+    }
+
+    private Void handleAsyncException(Throwable throwable) {
+        logger.info("MSG: " + throwable.getMessage().toString());
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: You would normally want all these logs in one. You can easily concatenate them using the `{}` placeholders I described above. Additionally, you shouldn't need to print the stack trace etc manually. You could just try something like 

`logger.warn("Exception occurred:", throwable);`

Also note that I've used `warn` and not `info` as the level. You could also argue that you want to use `error` instead.

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java [hunk 11]
```diff
@@ -33,6 +62,7 @@ public void getTweetsAndSaveToDB() {
             SavedTweet savedTweet = new SavedTweet(tweet.getId(), tweet.getText(), "-1");
             tweetRepository.insert(savedTweet);
         }
-    }
 
+        logger.info("Tweets saved at " + new Date());
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: Same here, use `{}` placeholders

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TwittercrawlerApplication.java [hunk 12]
```diff
@@ -1,14 +1,18 @@
 package amcmurray.bw.twittercrawler;
 
 import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.autoconfigure.social.TwitterAutoConfiguration;
+import org.springframework.scheduling.annotation.EnableScheduling;
 
+@EnableScheduling
+@EnableAutoConfiguration(exclude = {TwitterAutoConfiguration.class})
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TwittercrawlerApplication.java]: Are the curly brackets required?

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java [hunk 13]
```diff
@@ -9,30 +17,51 @@
...
         this.twitter = twitter;
+        this.scheduledTaskExecutorService = scheduledTaskExecutorService;
     }
 
+    //cron set to every 15 minutes on the hour, eg 12:00, 12:15 etc
+    @Scheduled(cron = "0 0/15 * * * *")
     public void getTweetsAndSaveToDB() {
 
+        logger.info("Scheduled task started at {}", LocalDateTime.now().format(formatter));
+        CompletableFuture.runAsync(() -> saveRandomTweetsIntoDB(),
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: I would do the line break like this:

```java
CompletableFuture.runAsync(() -> saveRandomTweetsIntoDB(), scheduledTaskExecutorService)
        .exceptionally(throwable -> handleAsyncException(throwable));
```

or

```java
CompletableFuture
        .runAsync(() -> saveRandomTweetsIntoDB(), scheduledTaskExecutorService)
        .exceptionally(throwable -> handleAsyncException(throwable));
```

which should make it much more readable

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java [hunk 14]
```diff
@@ -9,30 +17,51 @@
...
+                scheduledTaskExecutorService).exceptionally(throwable -> handleAsyncException(throwable));
+    }
+
+    private Void handleAsyncException(Throwable throwable) {
+        logger.warn("Exception occurred: ", throwable);
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: You could make this log more explicit and say `"Exception while saving mentions into database: "` or so

File: twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java [hunk 15]
```diff
@@ -9,30 +17,51 @@
...
+
+    private void saveRandomTweetsIntoDB() {
         //searching for a blank space in a tweet, in english
         SearchParameters params = new SearchParameters("%20");
         params.lang("en");
         SearchResults rawSearch = twitter.searchOperations().search(params);
 
         for (Tweet tweet : rawSearch.getTweets()) {
-            SavedTweet savedTweet = new SavedTweet(tweet.getId(), tweet.getText(), "-1");
+            SavedTweet savedTweet = new SavedTweet(tweet.getId(), tweet.getText(), -1);
```
Comment (YNedderhoff) [inline twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: I wouldn't give this the tweet id. The reason is that once you match these Tweets against your queries, you can have one `Tweet` generate multiple mentions (aka SavedTweets) because they match multiple queries (if this Tweet - Query - Mention relation isn't entirely clear we can talk about it again).

I haven't read this doc fully, but do you think one of the proposed solutions are suitable and doable for your usecase?
https://www.mongodb.com/blog/post/generating-globally-unique-identifiers-for-use-with-mongodb

Comment (YNedderhoff) [reply twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: Ah and one more thing. `Tweet` probably has more data, right, like date, maybe language, location, or so?

I'd store those in `SavedTweet`, then you can filter by them later.

Comment (amcmurray-bw) [reply twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: Would it be better to rename `SavedTweet` to something like `SavedMentionTwitter` or `TwitterMention`?

Comment (YNedderhoff) [reply twitter-crawler/src/main/java/amcmurray/bw/twittercrawler/TweetService.java]: If you compare it to analytics, there mentions are actually the same objects regardless of the data source. Just `Mention`. And they hold a `pageType` field which can have values like 'Twitter', 'Instagram', ...

So I'd suggest `Mention`
